File: server/src/main/java/touchflip/core/GameState.java
package touchflip.core;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/**
 * 게임 규칙의 유일한 소유자. 클라이언트 src/core/GameState.ts 의 1:1 포팅.
 *
 * 절대 규칙:
 * - 시계·IO·네트워크에 의존하지 않는다. 모든 시각은 인자로 주입받는다.
 * - 입력은 ServerInputGovernor 를 통과한 TapEvent 만 받는다.
 * - 스레드 안전하지 않다. Room 이 자기 락 안에서만 호출한다.
 */
public final class GameState {
    private final GameRules rules;

    private Phase phase = Phase.IDLE;
    private int[] board;
    private int round = 1;
    private final Map<PlayerId, Integer> roundWins = new EnumMap<>(PlayerId.class);
    private final Map<PlayerId, Boolean> ready = new EnumMap<>(PlayerId.class);
    private final Map<PlayerId, Integer> fouls = new EnumMap<>(PlayerId.class);
    private final Map<PlayerId, Long> penaltyUntil = new EnumMap<>(PlayerId.class);

    private long countdownEndsAt;
    private long roundStartAt;
    private long roundEndsAt;

    private RoundResult lastRound;
    private PlayerId matchWinner;
    private long revision;

    private final List<Consumer<GameEvent>> listeners = new ArrayList<>();

    public GameState(GameRules rules) {
        this.rules = rules;
        this.board = new int[rules.cardCount()];
        for (PlayerId p : PlayerId.values()) {
            roundWins.put(p, 0);
            ready.put(p, false);
            fouls.put(p, 0);
            penaltyUntil.put(p, 0L);
        }
    }

    // ---------------------------------------------------------------- 읽기

    public Phase phase() { return phase; }
    public int[] board() { return board.clone(); }
    public int round() { return round; }
    public int roundWins(PlayerId p) { return roundWins.get(p); }
    public int orientation() { return round % 2 == 1 ? 0 : 180; }
    public PlayerId matchWinner() { return matchWinner; }
    public long roundDurationMs() { return rules.roundDurationMs(); }
    public long revision() { return revision; }
    public boolean isReady(PlayerId p) { return ready.get(p); }
    public int fouls(PlayerId p) { return fouls.get(p); }
    public RoundResult lastRound() { return lastRound; }

    public int cardCount(PlayerId p) {
        int n = 0;
        for (int owner : board) if (owner == p.number) n++;
        return n;
    }

    public long penaltyUntil(PlayerId p) { return penaltyUntil.get(p); }

    public long remainingMs(long now) {
        return switch (phase) {
            case PLAYING -> Math.max(0, roundEndsAt - now);
            case COUNTDOWN -> Math.max(0, countdownEndsAt - now);
            case IDLE -> rules.roundDurationMs();
            default -> 0;
        };
    }

    public Snapshot snapshot(long now) {
        long remaining = remainingMs(now);
        List<Integer> boardList = new ArrayList<>(board.length);
        for (int owner : board) boardList.add(owner);
        return new Snapshot(
                phase.wire,
                boardList,
                round,
                rules.roundsToWin() * 2 - 1,
                orientation(),
                new int[] {cardCount(PlayerId.P1), cardCount(PlayerId.P2)},
                new int[] {roundWins.get(PlayerId.P1), roundWins.get(PlayerId.P2)},
                new boolean[] {ready.get(PlayerId.P1), ready.get(PlayerId.P2)},
                new int[] {fouls.get(PlayerId.P1), fouls.get(PlayerId.P2)},
                new long[] {penaltyUntil.get(PlayerId.P1), penaltyUntil.get(PlayerId.P2)},
                rules.roundDurationMs(),
                remaining,
                phase == Phase.COUNTDOWN ? (int) Math.max(1, Math.ceil(remaining / 1000.0)) : 0,
                Snapshot.RoundResultWire.of(lastRound),
                matchWinner == null ? null : matchWinner.number,
                revision);
    }

    // ---------------------------------------------------------------- 쓰기

    public Runnable subscribe(Consumer<GameEvent> listener) {
        listeners.add(listener);
        return () -> listeners.remove(listener);
    }

    /** 준비 홀드 상태 변경. 양쪽이 동시에 누르고 있어야 카운트다운, 하나라도 놓으면 취소. */
    public void setReady(PlayerId player, boolean held, long now) {
        if (ready.get(player) == held) return;
        ready.put(player, held);
        emit(new GameEvent.Ready(player, held));

        if (!held) {
            if (phase == Phase.COUNTDOWN) abortCountdown();
            bump();
            return;
        }

        if (ready.get(PlayerId.P1) && ready.get(PlayerId.P2)) {
            switch (phase) {
                case IDLE -> beginCountdown(now);
                case ROUND_END -> {
                    // 무승부 라운드는 승수를 주지 않았으므로 같은 라운드 번호로 재경기한다
                    if (lastRound != null && lastRound.winner() != null) round += 1;
                    beginCountdown(now);
                }
                case MATCH_END -> {
                    resetMatch();
                    ready.put(PlayerId.P1, true);
                    ready.put(PlayerId.P2, true);
                    beginCountdown(now);
                }
                default -> { }
            }
        }
        bump();
    }

    /**
     * 유효 탭 처리. 카운트다운 중 탭은 부정 출발로 기록하고 거부한다.
     * 라운드 종료 시각 이후의 timestamp 는 tick 이 아직 안 돌았더라도 거부한다.
     */
    public TapResult applyTap(TapEvent tap) {
        int i = tap.cardIndex();
        if (i < 0 || i >= board.length) {
            return new TapResult(false, i, tap.player(), 0, 0, "out-of-range");
        }
        int previousOwner = board[i];
        PlayerId p = tap.player();
        long t = tap.timestamp();

        if (phase == Phase.COUNTDOWN) {
            fouls.put(p, fouls.get(p) + 1);
            emit(new GameEvent.Foul(p, t));
            bump();
            return TapResult.reject(tap, previousOwner, "foul");
        }
        if (phase != Phase.PLAYING) return TapResult.reject(tap, previousOwner, "not-playing");
        if (t < roundStartAt || t >= roundEndsAt) return TapResult.reject(tap, previousOwner, "not-playing");
        if (t < penaltyUntil.get(p)) return TapResult.reject(tap, previousOwner, "penalty");
        if (previousOwner == p.number) return TapResult.reject(tap, previousOwner, "already-owned");

        board[i] = p.number;
        emit(new GameEvent.Flip(i, p.number, previousOwner, t));
        bump();
        return new TapResult(true, i, p, p.number, previousOwner, null);
    }

    /** 페이즈 전이. 매 틱 호출. */
    public void tick(long now) {
        if (phase == Phase.COUNTDOWN && now >= countdownEndsAt) {
            startRound();
            return;
        }
        if (phase == Phase.PLAYING && now >= roundEndsAt) endRound();
    }

    /** 서버 전용: 상대가 이탈했을 때 부전승 처리 */
    public void forfeit(PlayerId winner) {
        Phase previous = phase;
        matchWinner = winner;
        phase = Phase.MATCH_END;
        ready.put(PlayerId.P1, false);
        ready.put(PlayerId.P2, false);
        emit(new GameEvent.PhaseChange(Phase.MATCH_END, previous));
        emit(new GameEvent.MatchEnd(winner));
        bump();
    }

    public void resetMatch() {
        Phase previous = phase;
        phase = Phase.IDLE;
        board = new int[rules.cardCount()];
        round = 1;
        for (PlayerId p : PlayerId.values()) {
            roundWins.put(p, 0);
            ready.put(p, false);
            fouls.put(p, 0);
            penaltyUntil.put(p, 0L);
        }
        countdownEndsAt = roundStartAt = roundEndsAt = 0;
        lastRound = null;
        matchWinner = null;
        if (previous != Phase.IDLE) emit(new GameEvent.PhaseChange(Phase.IDLE, previous));
        bump();
    }

    // ---------------------------------------------------------------- 내부

    private void beginCountdown(long now) {
        Phase previous = phase;
        board = new int[rules.cardCount()];
        for (PlayerId p : PlayerId.values()) {
            fouls.put(p, 0);
            penaltyUntil.put(p, 0L);
        }
        countdownEndsAt = now + rules.countdownMs();
        phase = Phase.COUNTDOWN;
        emit(new GameEvent.PhaseChange(Phase.COUNTDOWN, previous));
    }

    private void abortCountdown() {
        Phase previous = phase;
        phase = Phase.IDLE;
        countdownEndsAt = 0;
        for (PlayerId p : PlayerId.values()) fouls.put(p, 0);
        emit(new GameEvent.PhaseChange(Phase.IDLE, previous));
    }

    private void startRound() {
        Phase previous = phase;
        // now 가 아니라 예정된 카운트다운 종료 시각 기준. 틱이 늦어도 라운드가 길어지지 않는다.
        roundStartAt = countdownEndsAt;
        roundEndsAt = roundStartAt + rules.roundDurationMs();
        phase = Phase.PLAYING;

        for (PlayerId p : PlayerId.values()) {
            int n = fouls.get(p);
            if (n <= 0) continue;
            long amount = rules.stackFalseStartPenalty() ? rules.falseStartPenaltyMs() * n : rules.falseStartPenaltyMs();
            penaltyUntil.put(p, roundStartAt + amount);
        }
        emit(new GameEvent.PhaseChange(Phase.PLAYING, previous));
        bump();
    }

    private void endRound() {
        Phase previous = phase;
        int c1 = cardCount(PlayerId.P1);
        int c2 = cardCount(PlayerId.P2);
        PlayerId winner = c1 == c2 ? null : (c1 > c2 ? PlayerId.P1 : PlayerId.P2);

        lastRound = new RoundResult(round, winner, c1, c2);
        ready.put(PlayerId.P1, false);
        ready.put(PlayerId.P2, false);
        if (winner != null) roundWins.put(winner, roundWins.get(winner) + 1);

        if (winner != null && roundWins.get(winner) >= rules.roundsToWin()) {
            matchWinner = winner;
            phase = Phase.MATCH_END;
            emit(new GameEvent.RoundEnd(lastRound));
            emit(new GameEvent.PhaseChange(Phase.MATCH_END, previous));
            emit(new GameEvent.MatchEnd(winner));
        } else {
            phase = Phase.ROUND_END;
            emit(new GameEvent.RoundEnd(lastRound));
            emit(new GameEvent.PhaseChange(Phase.ROUND_END, previous));
        }
        bump();
    }

    private void emit(GameEvent event) {
        for (Consumer<GameEvent> l : List.copyOf(listeners)) l.accept(event);
    }

    private void bump() { revision++; }

    @Override
    public String toString() {
        return "GameState{phase=" + phase + ", round=" + round + ", board=" + Arrays.toString(board) + "}";
    }
}

File: src/core/GameState.ts
import type { GameConfig } from '../config';
import type {
  CardOwner,
  GameEvent,
  MatchOptions,
  Phase,
  PlayerId,
  RoundResult,
  Snapshot,
  TapEvent,
  TapResult,
} from './types';

/**
 * 게임 규칙의 유일한 소유자.
 *
 * 절대 규칙:
 * - DOM 에 의존하지 않는다. document / window / performance 참조 금지.
 * - 시간을 스스로 읽지 않는다. 모든 시각은 인자로 주입받는다.
 *   (event.timeStamp 와 performance.now() 는 같은 time origin 이므로
 *    탭 판정 시각과 라운드 타이머가 같은 축 위에 놓인다.)
 * - 입력은 InputGovernor 를 통과한 TapEvent 만 받는다.
 */
export class GameState {
  private readonly cfg: GameConfig;
  /** 시작 화면에서 고른 라운드 길이. 없으면 config 기본값 */
  private readonly roundDuration: number;

  private _phase: Phase = 'idle';
  private _board: CardOwner[];
  private _round = 1;
  private _roundWins: Record<PlayerId, number> = { 1: 0, 2: 0 };
  private _ready: Record<PlayerId, boolean> = { 1: false, 2: false };
  private _fouls: Record<PlayerId, number> = { 1: 0, 2: 0 };
  private _penaltyUntil: Record<PlayerId, number> = { 1: 0, 2: 0 };

  private _countdownEndsAt = 0;
  private _roundStartAt = 0;
  private _roundEndsAt = 0;

  private _lastRound: RoundResult | null = null;
  private _matchWinner: PlayerId | null = null;
  private _revision = 0;

  private readonly listeners = new Set<(event: GameEvent) => void>();

  constructor(cfg: GameConfig, options: Partial<Pick<MatchOptions, 'roundDurationMs'>> = {}) {
    this.cfg = cfg;
    this.roundDuration = options.roundDurationMs ?? cfg.match.ROUND_DURATION_MS;
    this._board = new Array<CardOwner>(cfg.board.CARD_COUNT).fill(0);
  }

  // ---------------------------------------------------------------- 읽기

  get phase(): Phase {
    return this._phase;
  }

  get board(): readonly CardOwner[] {
    return this._board;
  }

  get round(): number {
    return this._round;
  }

  get roundWins(): Record<PlayerId, number> {
    return { ...this._roundWins };
  }

  get orientation(): 0 | 180 {
    // 홀수 라운드 0도, 짝수 라운드 180도
    return this._round % 2 === 1 ? 0 : 180;
  }

  get matchWinner(): PlayerId | null {
    return this._matchWinner;
  }

  get roundDurationMs(): number {
    return this.roundDuration;
  }

  cardCount(player: PlayerId): number {
    let n = 0;
    for (const owner of this._board) if (owner === player) n += 1;
    return n;
  }

  penaltyUntil(player: PlayerId): number {
    return this._penaltyUntil[player];
  }

  remainingMs(now: number): number {
    if (this._phase === 'playing') return Math.max(0, this._roundEndsAt - now);
    if (this._phase === 'countdown') return Math.max(0, this._countdownEndsAt - now);
    if (this._phase === 'idle') return this.roundDuration;
    return 0;
  }

  snapshot(now: number): Snapshot {
    const remainingMs = this.remainingMs(now);
    return {
      phase: this._phase,
      board: this._board,
      round: this._round,
      maxRounds: this.cfg.match.MAX_ROUNDS,
      orientation: this.orientation,
      counts: { 1: this.cardCount(1), 2: this.cardCount(2) },
      roundWins: { ...this._roundWins },
      ready: { ...this._ready },
      fouls: { ...this._fouls },
      penaltyUntil: { ...this._penaltyUntil },
      roundDurationMs: this.roundDuration,
      remainingMs,
      countdownValue: this._phase === 'countdown' ? Math.max(1, Math.ceil(remainingMs / 1000)) : 0,
      lastRound: this._lastRound,
      matchWinner: this._matchWinner,
      revision: this._revision,
    };
  }

  // ---------------------------------------------------------------- 쓰기

  subscribe(listener: (event: GameEvent) => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  /**
   * 준비 버튼 홀드 상태 변경.
   * 양쪽이 동시에 누르고 있어야 카운트다운이 시작되고, 도중에 하나라도 놓으면 취소된다.
   */
  setReady(player: PlayerId, held: boolean, now: number): void {
    if (this._ready[player] === held) return;
    this._ready[player] = held;
    this.emit({ type: 'ready', playerId: player, held });

    if (!held) {
      if (this._phase === 'countdown') this.abortCountdown();
      this.bump();
      return;
    }

    if (this._ready[1] && this._ready[2]) {
      switch (this._phase) {
        case 'idle':
          this.beginCountdown(now);
          break;
        case 'roundEnd':
          // 무승부 라운드는 승수를 주지 않았으므로 같은 라운드 번호로 재경기한다
          if (this._lastRound?.winner != null) this._round += 1;
          this.beginCountdown(now);
          break;
        case 'matchEnd':
          this.resetMatch(now);
          this._ready = { 1: true, 2: true };
          this.beginCountdown(now);
          break;
        default:
          break;
      }
    }
    this.bump();
  }

  /**
   * 유효 탭 처리. InputGovernor 를 통과한 이벤트만 들어온다.
   *
   * 카운트다운 중의 탭은 거부하고 부정 출발로 기록한다.
   * 라운드 종료 시각 이후의 timestamp 는 tick 이 아직 안 돌았더라도 거부한다.
   * (rAF 주기에 따라 판정이 달라지면 주사율 높은 기기가 유리해진다)
   */
  applyTap(tap: TapEvent): TapResult {
    const { cardIndex, playerId, timestamp } = tap;
    const previousOwner = this._board[cardIndex];

    if (previousOwner === undefined) {
      return {
        accepted: false,
        cardIndex,
        playerId,
        owner: 0,
        previousOwner: 0,
        reason: 'out-of-range',
      };
    }

    if (this._phase === 'countdown') {
      this._fouls[playerId] += 1;
      this.emit({ type: 'foul', playerId, timestamp });
      this.bump();
      return this.reject(tap, previousOwner, 'foul');
    }

    if (this._phase !== 'playing') return this.reject(tap, previousOwner, 'not-playing');
    if (timestamp < this._roundStartAt || timestamp >= this._roundEndsAt) {
      return this.reject(tap, previousOwner, 'not-playing');
    }
    if (timestamp < this._penaltyUntil[playerId]) return this.reject(tap, previousOwner, 'penalty');
    if (previousOwner === playerId) return this.reject(tap, previousOwner, 'already-owned');

    // 상태는 즉시 반영된다. 애니메이션은 렌더러가 알아서 하며 여기를 블로킹하지 않는다.
    this._board[cardIndex] = playerId;
    this.emit({ type: 'flip', cardIndex, owner: playerId, previousOwner, timestamp });
    this.bump();

    return { accepted: true, cardIndex, playerId, owner: playerId, previousOwner };
  }

  /** 페이즈 전이를 구동한다. 매 프레임 호출해도 되고, 테스트에서 임의 시각으로 호출해도 된다. */
  tick(now: number): void {
    if (this._phase === 'countdown' && now >= this._countdownEndsAt) {
      this.startRound();
      return;
    }
    if (this._phase === 'playing' && now >= this._roundEndsAt) {
      this.endRound();
    }
  }

  /** 매치 전체 초기화 */
  resetMatch(_now: number): void {
    const previous = this._phase;
    this._phase = 'idle';
    this._board = new Array<CardOwner>(this.cfg.board.CARD_COUNT).fill(0);
    this._round = 1;
    this._roundWins = { 1: 0, 2: 0 };
    this._ready = { 1: false, 2: false };
    this._fouls = { 1: 0, 2: 0 };
    this._penaltyUntil = { 1: 0, 2: 0 };
    this._countdownEndsAt = 0;
    this._roundStartAt = 0;
    this._roundEndsAt = 0;
    this._lastRound = null;
    this._matchWinner = null;
    if (previous !== 'idle') this.emit({ type: 'phase', phase: 'idle', previous });
    this.bump();
  }

  // ---------------------------------------------------------------- 내부

  private beginCountdown(now: number): void {
    const previous = this._phase;
    this._board = new Array<CardOwner>(this.cfg.board.CARD_COUNT).fill(0);
    this._fouls = { 1: 0, 2: 0 };
    this._penaltyUntil = { 1: 0, 2: 0 };
    this._countdownEndsAt = now + this.cfg.match.COUNTDOWN_MS;
    this._phase = 'countdown';
    this.emit({ type: 'phase', phase: 'countdown', previous });
  }

  private abortCountdown(): void {
    const previous = this._phase;
    this._phase = 'idle';
    this._countdownEndsAt = 0;
    this._fouls = { 1: 0, 2: 0 };
    this.emit({ type: 'phase', phase: 'idle', previous });
  }

  private startRound(): void {
    const previous = this._phase;
    // now 가 아니라 예정된 카운트다운 종료 시각을 기준으로 삼는다.
    // rAF 호출이 늦은 기기에서 라운드가 길어지는 것을 막는다.
    this._roundStartAt = this._countdownEndsAt;
    this._roundEndsAt = this._roundStartAt + this.roundDuration;
    this._phase = 'playing';

    const penalty = this.cfg.match.FALSE_START_PENALTY_MS;
    for (const player of [1, 2] as PlayerId[]) {
      const fouls = this._fouls[player];
      if (fouls <= 0) continue;
      const amount = this.cfg.match.STACK_FALSE_START_PENALTY ? penalty * fouls : penalty;
      // 페널티는 라운드 시작 시점부터 적용된다. 카운트다운 중에 정지시켜봐야 의미가 없다.
      this._penaltyUntil[player] = this._roundStartAt + amount;
    }

    this.emit({ type: 'phase', phase: 'playing', previous });
    this.bump();
  }

  private endRound(): void {
    const previous = this._phase;
    const counts: Record<PlayerId, number> = { 1: this.cardCount(1), 2: this.cardCount(2) };
    const winner: PlayerId | null = counts[1] === counts[2] ? null : counts[1] > counts[2] ? 1 : 2;

    const result: RoundResult = { round: this._round, winner, counts };
    this._lastRound = result;
    this._ready = { 1: false, 2: false };

    if (winner != null) this._roundWins[winner] += 1;

    if (winner != null && this._roundWins[winner] >= this.cfg.match.ROUNDS_TO_WIN) {
      this._matchWinner = winner;
      this._phase = 'matchEnd';
      this.emit({ type: 'roundEnd', result });
      this.emit({ type: 'phase', phase: 'matchEnd', previous });
      this.emit({ type: 'matchEnd', winner });
    } else {
      this._phase = 'roundEnd';
      this.emit({ type: 'roundEnd', result });
      this.emit({ type: 'phase', phase: 'roundEnd', previous });
    }
    this.bump();
  }

  private reject(tap: TapEvent, previousOwner: CardOwner, reason: TapResult['reason']): TapResult {
    return {
      accepted: false,
      cardIndex: tap.cardIndex,
      playerId: tap.playerId,
      owner: previousOwner,
      previousOwner,
      ...(reason ? { reason } : {}),
    };
  }

  private emit(event: GameEvent): void {
    for (const listener of this.listeners) listener(event);
  }

  private bump(): void {
    this._revision += 1;
  }
}
